openembedded-core-mbl/scripts/lib/mbl-signing/signing/pki/x509utils.py

```diff
@@ -0,0 +1,315 @@
+
+
+"""X509 extension and certificate objects."""
+
+import binascii
+from abc import abstractmethod
+
+from cryptography import x509
+from cryptography import utils
+from cryptography.hazmat.primitives import hashes
+from cryptography.hazmat.primitives import serialization
+from cryptography.hazmat.backends import default_backend
+
+from signing.pki import asn1parse
+
+
+@utils.register_interface(x509.ExtensionType)
+class ATFExtension:
+    """Describes a custom X509v3 extension used by ATF."""
+
+    def __init__(self, oid, critical=False, value=None):
+        """
+        Initialise the extension object.
+
+        :params oid string: Dotted string OID for the extension type.
+        :params critical bool: Specifies whether to mark this
+        extension as critical or not.
+        :params value bytes: Extension value in ASN1 format.
+        """
+        self._oid = x509.ObjectIdentifier(oid)
+        self._critical = critical
+        self._value = x509.UnrecognizedExtension(self.oid, value)
+
+    def __str__(self):
+        """Return a string representation of the object."""
+        return "{};{};{}".format(
+            self.oid.dotted_string,
+            str(self.critical).lower(),
+            binascii.hexlify(self.value.value).decode(),
+        )
+
+    @property
+    def oid(self):
+        """Object Identifier."""
+        return self._oid
+
+    @property
+    def critical(self):
+        """Specify if the extension will be marked as critical."""
+        return self._critical
+
+    @property
+    def value(self):
+        """Extension value wrapped in an UnrecognizedExtension."""
+        return self._value
+
+    @value.setter
+    def value(self, new_val):
+        """
+        Extension value setter.
+
+        :params new_val bytes: New value for the extension.
+        """
+        self._value = x509.UnrecognizedExtension(self.oid, new_val)
+
+    @abstractmethod
+    def encode_value(self, value_data):
+        """Subclass must override this to encode specific value types."""
+        pass
+
+
+class HashExtension(ATFExtension):
+    """An ATF extension containing a hash of a boot component."""
+
+    def __init__(self, *args, content_file=None, **kwargs):
+        """
+        Initialise the extension object.
+
+        :params content_file str: Name of the content file to hash.
+        """
+        super().__init__(*args, **kwargs)
+        self.content_file = content_file
+
+    def encode_value(self, data):
+        """
+        Create a digest of content_file and encode to an ASN1 value.
+
+        :params data bytes: File data in bytes.
+        """
+        if data is None:
+            digest = b"\x00" * 32
+        else:
+            hasher = hashes.Hash(hashes.SHA256(), backend=default_backend())
+            hasher.update(data)
+            digest = hasher.finalize()
+        enc_val = asn1parse.encode_asn1_digest_info(
+            "2.16.840.1.101.3.4.2.1", digest
```

Comment:
Where does "2.16.840.1.101.3.4.2.1" come from?

Reply:
Oh I found it in the OID repository - http://oid-info.com/get/2.16.840.1.101.3.4.2.1